--- frama-c-Nitrogen-20111001.old/configure.in	2011-10-10 09:40:09.000000000 +0100
+++ frama-c-Nitrogen-20111001/configure.in	2012-07-30 09:12:57.276551927 +0100
@@ -118,7 +118,7 @@
          fi
          ;;
   3.10*) echo "${ECHO_T}good!";;
-  3.1*) echo "${ECHO_T}good!"; OCAML_ANNOT_OPTION="-annot";;
+  3.1*|4*) echo "${ECHO_T}good!"; OCAML_ANNOT_OPTION="-annot";;
   *) echo "${ECHO_T}Incompatible version!"; exit 2;;
 esac
 
