--- src/plugins/wp/ProverWhy3.ml.orig	2019-12-03 17:00:00.000000000 -0700
+++ src/plugins/wp/ProverWhy3.ml	2020-03-24 20:52:00.143904622 -0600
@@ -128,14 +128,14 @@ let t_app' ~cnv ~f ~l ~p tl ty =
 *)
 
 let const_int (z:Z.t) =
-  Why3.(Term.t_const Number.(const_of_big_int (BigInt.of_string (Z.to_string z)))) Why3.Ty.ty_int
+  Why3.(Term.t_const (Constant.(ConstInt Number.{ il_kind = ILitDec; il_int = BigInt.of_string (Z.to_string z) }))) Why3.Ty.ty_int
 
 let const_real ~cnv (q:Q.t) =
   let mk_real_int z =
     let rc_negative = Z.sign z < 0 in
     let z = Z.abs z in
-    let rc_abs = Why3.Number.real_const_dec (Z.to_string z) "" None in
-    let c = Why3.Number.ConstReal { Why3.Number.rc_negative; rc_abs } in
+    let rc_abs = Why3.Number.real_literal ~radix:10 ~neg:rc_negative ~int:(Z.to_string z) ~frac:"" ~exp:None in
+    let c = Why3.Constant.ConstReal rc_abs in
     Why3.(Term.t_const c) Why3.Ty.ty_real
   in
   if Z.equal Z.one q.den
@@ -1089,7 +1089,7 @@ let ping_prover_call p =
       Wp_parameters.debug ~dkey
         "@[@[Why3 result for %a:@] @[%a@] and @[%a@]@."
         Why3.Whyconf.print_prover p.prover
-        (Why3.Call_provers.print_prover_result) pr
+        (Why3.Call_provers.print_prover_result ~json_model:false) pr
         VCS.pp_result r;
       Task.Return (Task.Result r)
 
